server/CGameHandler.cpp

```diff
@@ -3432,6 +3434,17 @@ bool CGameHandler::isAllowedExchange(ObjectInstanceID id1, ObjectInstanceID id2)
 				return true;
 		}
 
+		if(o1->ID == Obj::ALTAR_OF_SACRIFICE)
+		{
+			return true;
+		}
+		if(o2->ID == Obj::ALTAR_OF_SACRIFICE)
+		{
+			const auto visitingHero = getVisitingHero(o2);
+			const auto thisHero = static_cast<const CGHeroInstance*>(o1);
+			if(visitingHero == thisHero)
+				return true;
+		}
```

Comment:
Instead of checking for object ID you should check that:
1) Object can be casted to IMarket
2) If yes - that object allows trade using EMarketMode::ARTIFACT_EXP mode

Reply:
In ApplyGhNetPackVisitor::visitBulkExchangeArtifacts should also be type cast instead of ID comparsion?

Reply:
done